### ISSUE TITLE
Building .debs from .qt5 release tarball on Debian 

### ISSUE THREAD

**Issue by alexolivan** [ElvishArtisan/rivendell#688]
HI!

I'm starting to prepare to switch to qt5 branch to test it... now but, from the time I early tested qt5, a huuuge amount of work happened there: now there's a nice debian folder, and we got a debianish-ready tarball!!! So, I want to stick to packages generated with official debian packaging, and drop from custom/weird packaging :-)

My first attempt has been building rivendell_4.0.0beta1 in Debian Bullseye (still testing), and it ALMOST work out of the box :-D
Here's but the two small changes I needed to do to get my first 'officialish'  Debian Packages ... please consider the feasibility of adding those (or something equivalent) to achieve full out-of-the-box experience:

1. debian/control
The file states the dependency on hpklinux-dev package... but this package is not provided (AFAIK) neither from AudioScience PPA nor from driver pbuild scripts... instead, libhpi-dev is the actual package naming... so I appended ' | libhpi-dev' so either of both satisfy the dependency: 
 
```
Source: rivendell
Section: audio
Priority: optional
Maintainer: Fred Gleason <fredg@paravelsystems.com>
Build-Depends: debhelper-compat (= 12), autotools-dev, hpklinux-dev | libhpi-dev
Standards-Version: 4.4.1
```

2. debian/rules 
Build fails due to the known environment DOCBOOK_STYLESHEETS need ... and that is solved easily by exporting that at the top of the rules file. Please consider the possibility to add that to the file so it build out-of-the-box

```
#!/usr/bin/make -f

export DOCBOOK_STYLESHEETS=/usr/share/xml/docbook/stylesheet/docbook-xsl-ns

```

With these two changes, dpkg-buildpackage -us -uc generates the packages.

Looking forward for beta2 release to package that and test those nice GUI enhancements! :-)
Thank you very much and best regards!

**Comment by edgeradio993fm** [comment]
Good pointers. 

I had some fun getting the packages to build under Debian 10. Using the build_debs.sh script was very hit and miss. Mostly miss. I have more luck using debuild. Nice to have docbook working. Seems like the environment variable were being ignored during compile. The --sanitize-env switch wasn't a good solution.

**Comment by alexolivan** [comment]
Yep... building custom packages was tricky, quite a lot... now but, with an officially maintained debian folder and Debian-packaging-friendly code it is waaaay easier, and the resulting packages wil be consistent among users.

Moreover, now the door opens wide for what could be a milestone in Rivendell's users-community: a Debian RFP. If a Debian RFP will materialize into a Debian (Repo) inclusion, then, Rivendell will soon appear available in all Debian-based distros (Ubuntu, Mint, etc...) ... that would make it the reference, professional, full-featured and opensource radio-automation software (IMHO Rivendell deserves that... i think of it as the equivalent on radio automation to Libreoffice for office suite)

 Best regards!

**Comment by ElvishArtisan** [comment]
> The file states the dependency on hpklinux-dev package... but this package is not provided (AFAIK) neither from AudioScience PPA nor from driver pbuild scripts... instead, libhpi-dev is the actual package naming... so I appended ' | libhpi-dev' so either of both satisfy the dependency.

Ok, here's the deal with AudioScience HPI support on Linux: historically, various distros have shipped various versions of the AudioScience driver. Most of them don't enable HPI support at all (Debian Bullseye appears to the major exception to this; those 'libhpi' packages don't exist in Ubuntu 20.02 at all). And, the version of AudioScience driver shipped often significantly lags the current version provided by AudioScience; a real problem when ASI introduces a new card model and then users find that they cannot use it because the old driver doesn't recognize it. OR, _equally_ a problem, the shipped driver is _too new_, meaning that it doesn't recognize the older, "end-of_lifed" hardware whose support has been removed from the newer driver versions.

Hence, from very early on, the Rivendell project's practice has been to provide its own HPI driver RPMs alongside the Rivendell ones. These drivers are configured to provide HPI support _only_, and typically include a range of versions (including older versions that contain support for various 'end-of-lifed' models that are no longer supported by the 'latest-and-greatest' version from ASI). This allows users to pick the specific version they need that will work with whatever hardware they have. Unfortunately, this setup coexists very poorly with most distro-sbased software distrbution methods, which generally operate under the assumption that the latest version is _always_ the preferred one to use.

So, for the debian-ish ecosystem, we've extended that approach by building 'hpklinux' packages (the same base name used for the RPM versions) that contain the HPI driver. They are currently available online for Ubuntu 20.02 in the same repo as the Rivendell packages: https://software.paravelsystems.com/ubuntu/dists/focal/main/. The 'debian/' data for generating the packages can be found at: https://github.com/ElvishArtisan/debian-hpklinux. Feel free to submit PRs there to address any issues found with building packages under other Debian-derived distros.

As for distro-native 'libhpi' packages, I'd really prefer to avoid them, for the reasons stated above. You are of course free to do whatever you wish with Rivendell within the confines of GPLv2, but I'd rather avoid a situation where we have sets of competing DEB packages in the wild that have different dependencies. That's liable to make for a very confusing experience for users, most of whom after all are looking for an automation solution that 'just works' out of the box.

**Comment by ElvishArtisan** [comment]
> Build fails due to the known environment DOCBOOK_STYLESHEETS need ... and that is solved easily by exporting that at the top of the rules file. Please consider the possibility to add that to the file so it build out-of-the-box.

Take a look at "repo/install_rivendell_devel.sh" in https://github.com/ElvishArtisan/ubuntu-rivendell-installer. This will install and configure the full dev environment needed for Rivendell v4.x on debian-ish systems, including any needed environmental variables. (Currently only tested on Ubuntu 20.02, but should be very close to whats needed on recent Debian versions).

**Comment by ElvishArtisan** [comment]
> Moreover, now the door opens wide for what could be a milestone in Rivendell's users-community: a Debian RFP. If a Debian RFP will materialize into a Debian (Repo) inclusion, then, Rivendell will soon appear available in all Debian-based distros (Ubuntu, Mint, etc...) ... that would make it the reference, professional, full-featured and opensource radio-automation software (IMHO Rivendell deserves that... i think of it as the equivalent on radio automation to Libreoffice for office suite)

I actually attempted to file an ITP for Rivendell with Debian a few months back. It basically disappeared into the memory hole; I never did see it appear in the tracking system or get any feedback about it whatsoever. And, I've really been too busy coding to deal with the incredibly byzantine process bureaucracy that surrounds Debian.

**Comment by alexolivan** [comment]
> Take a look at "repo/install_rivendell_devel.sh" in https://github.com/ElvishArtisan/ubuntu-rivendell-installer. This will install and configure the full dev environment needed for Rivendell v4.x on debian-ish systems, including any needed environmental variables. (Currently only tested on Ubuntu 20.02, but should be very close to whats needed on recent Debian versions).

Very interesting... will keep an eye!

**Comment by alexolivan** [comment]
> Ok, here's the deal with AudioScience HPI support on Linux: historically, various distros have shipped various versions of the AudioScience driver. Most of them don't enable HPI support at all (Debian Bullseye appears to the major exception to this; those 'libhpi' packages don't exist in Ubuntu 20.02 at all). And, the version of AudioScience driver shipped often significantly lags the current version provided by AudioScience; a real problem when ASI introduces a new card model and then users find that they cannot use it because the old driver doesn't recognize it. OR, _equally_ a problem, the shipped driver is _too new_, meaning that it doesn't recognize the older, "end-of_lifed" hardware whose support has been removed from the newer driver versions.

Hi, The thing is that AudioScience has indeed an Ubuntu repository for the official latest driver (Delio Brignoli is AudioScience staff maintainer... he helped me a lot!) .... here's Ubuntu/PPA link to the packaged contents (I miss Focal Fossa there now): `https://launchpad.net/~audioscience/+archive/ubuntu/release/+packages` 

This is an Ubuntu repository and I was reluctant at first to setup it in Buster... but as Delio pointed me time ago, it works on Debian (tested in Buster). The repo .deb packages are the same that are generated using the scripts (based pbuilder... I didn't liked that since it install a huuuuge amount of stuff) provided in the Linux drivers tarballs... that's why I finally gave up and started using that Launchpad PPA.

Anyways, the thing is that, as you mention, looking there, there is a (source I guess) package named hpklinux-2.20.33~XXXXX (I use bionic3 for XXXXX for Buster) but it builds into libhpi, libhpi-dev and libhpiudp binary packages for i386 and amd64 archs

And here's an example of how the official PPA stuff looks like once setup in a Debian buster system:
```
root@debian:~# apt-cache policy libhpi-dev
libhpi-dev:
  Installed: (none)
  Candidate: 4.20.33~bionic3
  Version table:
     4.20.33~bionic3 500
        500 http://ppa.launchpad.net/audioscience/release/ubuntu bionic/main amd64 Packages
```

I see but no Focal Fossa stuff there (I'm gonna ask Delio...) which is the one I bet will work in Bullseye. But, for Bionic/Buster and previous versions, I though it would be nice to be able to use the Ubuntu Official repo out-of-the-box.

Anyways... Thank you very much for the job!

**Comment by maxtimbo** [comment]
Can ya'll provide instructions for building on debian? I'll be happy to join in this quest of building for Debian...

**Comment by ElvishArtisan** [comment]
> Can ya'll provide instructions for building on debian? I'll be happy to join in this quest of building for Debian...

It's a production in Rivendell's build system. First, make sure you've got the required packages; see the "DISTRO-SPECIFIC NOTES" section in 'INSTALL' for some pointers.

Once you've installed those, the basic sequence (assuming a freshly checked-out source tree) is:

./autogen.sh
./configure
make deb

Mind you, the current setup has been tested only on Ubuntu 20.02, but I suspect that that's very close to what's needed on Debian Bullseye.

**Comment by alexolivan** [comment]
> The 'debian/' data for generating the packages can be found at: https://github.com/ElvishArtisan/debian-hpklinux. Feel free to submit PRs there to address any issues found with building packages under other Debian-derived distros.

Tried that... very interesting approach! ... however, as I feel you expected, it is somehow very Ubuntu specific:

```
root@buster:~# dpkg -i /home/alex/repos/hpklinux-dev_4.20.35-1_amd64.deb
(Reading database ... 79949 files and directories currently installed.)
Preparing to unpack .../hpklinux-dev_4.20.35-1_amd64.deb ...
Unpacking hpklinux-dev (4.20.35-1) over (4.20.35-1) ...
dpkg: dependency problems prevent configuration of hpklinux-dev:
 hpklinux-dev depends on hpklinux (= 4.20.35-1); however:
  Package hpklinux is not configured yet.

dpkg: error processing package hpklinux-dev (--install):
 dependency problems - leaving unconfigured
Errors were encountered while processing:
 hpklinux-dev
root@buster-builds:~# dpkg -i /home/administrator/repos/audioscience/hpklinux_4.20.35-1_amd64.deb
(Reading database ... 79949 files and directories currently installed.)
Preparing to unpack .../hpklinux_4.20.35-1_amd64.deb ...
Unpacking hpklinux (4.20.35-1) over (4.20.35-1) ...
dpkg: dependency problems prevent configuration of hpklinux:
 hpklinux depends on p7zip-full; however:
  Package p7zip-full is not installed.
 hpklinux depends on linux-headers-5.8.0-53-lowlatency; however:
  Package linux-headers-5.8.0-53-lowlatency is not installed.
 hpklinux depends on ubuntu-dev-tools; however:
  Package ubuntu-dev-tools is not installed.

dpkg: error processing package hpklinux (--install):
 dependency problems - leaving unconfigured
Processing triggers for libc-bin (2.28-10) ...
Errors were encountered while processing:
 hpklinux
```

What is more interesting to me is that it requires a very specific/matching kernel.


Best regards.

**Comment by ElvishArtisan** [comment]
> What is more interesting to me is that it requires a very specific/matching kernel.

Actually no -- not a specific kernel, but a specific set of kernel *headers*. Those are needed by [DKMS](https://en.wikipedia.org/wiki/Dynamic_Kernel_Module_Support) so it can build the appropriately versioned kernel module(s) after a kernel update.

Commit14ef428 in https://github.com/ElvishArtisan/debian-hpklinux modifies things so ```prepare_build.sh``` will generate the appropriate run-time dependency automatically. I'm moderately confident that this will now work just fine on recent Debians.

**Comment by alexolivan** [comment]
OK... I spoted the commit.I don't have AudioScience stuff around here... Will try ASAP and give feedback!
